rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,713 @@
+allowed, we can just rename A to B. We could also check that schemas don't remove fields unless
+those fields were already marked as deprecated, being able to enforce our unofficial guideline of
+not removing fields unless they've been marked as deprecated for at least one release.
+
+#### 4. Utilize `schemars` to generate the actual JSONSchema
+
+With the necessary information being derived from our configuration types directly, we still need a
+way to take that and emit an actual JSONSchema document for Vector's configuration. We would utilize
+[`schemars`][schemars], a crate for generating JSONSchema documents, for this purpose.
+
+The `schemars` crate, among other things, provides a small set of traits and types related
+specifically to programmatically generating a JSONSchema document. Types wish a user wishes to
+document must implement `JSONSchema`, which interacts with a `SchemaGenerator` object that actually
+holds the in-progress schema as a type is being walked.
+
+Our `Configurable` derive macro would implement this trait automatically for the given type, using
+the information provided by `Configurable` to ultimately drive the schema generation.
+
+## Rationale
+
+In general, the lack of a configuration schema, and being able to treat our Rust code as the source
+of truth, hurts both users and developers. As such, doing this work represents a huge opportunity to
+reduce developer toil when it comes to generating and synchronizing our generated documentation.  It
+additionally represents a large quality of life improvement for our users, who look to our
+documentation to be up-to-date and semantically meaningful.
+
+If we didn't do this work, we would be stuck with the current manual toil, which is not only extra
+work on the part of developers -- learning Cue, remembering to update it, etc -- but also reviewers,
+who now need to catch when these things are missed.
+
+Utilizing an extensible system for expressing the schema of Vector's configuration gives us the
+runway to solve our current problems, but also the ability to handle future problems and
+requirements as they come up.
+
+## Drawbacks
+
+The primary drawback of this approach is that it limits us to things we can reasonably express
+within the limitations of the Rust language itself, and what we're comfortable with representing via
+attributes. For example, single-line and multi-line doc comments are trivially supported, but if we
+wanted to start pushing more semantically-relevant information, such as configuration snippets, etc,
+it could be technically possible but appear as a very ugly attribute usage, which could make the
+source muddy and unclear.
+
+Additionally, while derive macros are written in Rust code, which can be much easier to grok than
+declarative macros, it represents a section of code that may be harder to understand and modify than
+if we used a more brute force approach.
```

Comment:
I was going to bring this up. Specifically, I often find myself going to https://serde.rs to find a specific attribute I want to add to my code. If we are going to rely heavily on the `Configurable` derive macro, I'd like for there to be a solution similar to the mentioned Serde website. For example, given the proc-macro has to live in a separate crate, we could add an extensive README to that crate which documents all available attributes.

Reply:
Yeah, we'd definitely end up with a large and comprehensive README.  No other way around it, given the lack of exploring documentation otherwise.

Reply:
Addressed in 7496d1aba.